### ISSUE TITLE
How to schedule gpus in volcano mpi job ?

### ISSUE THREAD

**Issue by nicklhy** [volcano-sh/volcano#474]
Currently, I am able to launch the hello-world job with mpi-example.yaml successfully. But as I mentioned [here](https://github.com/volcano-sh/volcano/issues/452#issuecomment-535851592), once I add the lines below following the official [tutorial](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/), the job stays in pending state forever.
```
resources:
  limits:
    nvidia.com/gpu: 1
```
The log of volcano-scheduler is pasted below
```
$ kubectl logs -n volcano-system volcano-scheduler-545cb4fb-8wfxf --tail 30
I1009 06:42:18.755838       1 scheduler_helper.go:49] Predicates failed for task <default/lm-mpi-job-mpiworker-0> on node <p40-3>: task default/lm-mpi-job-mpiworker-0 on node p40-3 fit failed: node(s) resource fit failed
I1009 06:42:18.755730       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-0> on node <p40-4>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 54020.00, memory 132631269376.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00, nvidia.com/gpu 0.00>
I1009 06:42:18.755856       1 scheduler_helper.go:49] Predicates failed for task <default/lm-mpi-job-mpiworker-0> on node <p40-4>: task default/lm-mpi-job-mpiworker-0 on node p40-4 fit failed: node(s) resource fit failed
I1009 06:42:18.755985       1 allocate.go:208] Binding Task <default/lm-mpi-job-mpiworker-0> to node <titanxp-4>
I1009 06:42:18.756036       1 statement.go:226] After allocated Task <default/lm-mpi-job-mpiworker-0> to Node <titanxp-4>: idle <cpu 31650.00, memory 270108176384.00, nvidia.com/gpu 7000.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00>, used <cpu 350.00, memory 31457280.00, nvidia.com/gpu 1000.00>, releasing <cpu 0.00, memory 0.00>
I1009 06:42:18.756159       1 statement.go:244] Allocating operations ...
I1009 06:42:18.756186       1 allocate.go:186] There are <7> nodes for Job <default/lm-mpi-job>
I1009 06:42:18.756224       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <vm10-1-0-10.ksc.com>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 7000.00, memory 16574566400.00, hugepages-2Mi 0.00>
I1009 06:42:18.756264       1 scheduler_helper.go:49] Predicates failed for task <default/lm-mpi-job-mpiworker-1> on node <vm10-1-0-10.ksc.com>: task default/lm-mpi-job-mpiworker-1 on node vm10-1-0-10.ksc.com fit failed: node(s) resource fit failed
I1009 06:42:18.756275       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <titanxp-4>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 31650.00, memory 270108176384.00, hugepages-2Mi 0.00, nvidia.com/gpu 7000.00, hugepages-1Gi 0.00>
I1009 06:42:18.756276       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <titanv-1>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 55750.00, memory 270135402496.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00, nvidia.com/gpu 8000.00>
I1009 06:42:18.756290       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <titanxp-1>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 31650.00, memory 270108180480.00, nvidia.com/gpu 8000.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00>
I1009 06:42:18.756311       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <p40-4>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 54020.00, memory 132631269376.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00, nvidia.com/gpu 0.00>
I1009 06:42:18.756272       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <p40-3>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 55610.00, memory 269944328192.00, hugepages-2Mi 0.00, nvidia.com/gpu 0.00, hugepages-1Gi 0.00>
I1009 06:42:18.756379       1 scheduler_helper.go:49] Predicates failed for task <default/lm-mpi-job-mpiworker-1> on node <p40-4>: task default/lm-mpi-job-mpiworker-1 on node p40-4 fit failed: node(s) resource fit failed
I1009 06:42:18.756305       1 scheduler_helper.go:44] Considering Task <default/lm-mpi-job-mpiworker-1> on node <p40-2>: <cpu 0.00, memory 0.00, nvidia.com/gpu 1000.00> vs. <cpu 55530.00, memory 134422245376.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00, nvidia.com/gpu 3000.00>
I1009 06:42:18.756388       1 scheduler_helper.go:49] Predicates failed for task <default/lm-mpi-job-mpiworker-1> on node <p40-3>: task default/lm-mpi-job-mpiworker-1 on node p40-3 fit failed: node(s) resource fit failed
I1009 06:42:18.756589       1 allocate.go:208] Binding Task <default/lm-mpi-job-mpiworker-1> to node <titanxp-1>
I1009 06:42:18.756641       1 statement.go:226] After allocated Task <default/lm-mpi-job-mpiworker-1> to Node <titanxp-1>: idle <cpu 31650.00, memory 270108180480.00, nvidia.com/gpu 7000.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00>, used <cpu 350.00, memory 31457280.00, nvidia.com/gpu 1000.00>, releasing <cpu 0.00, memory 0.00>
I1009 06:42:18.756787       1 statement.go:244] Allocating operations ...
I1009 06:42:18.756814       1 statement.go:310] Discarding operations ...
I1009 06:42:18.756832       1 statement.go:294] Remove Task <lm-mpi-job-mpiworker-1> on node <titanxp-1>
I1009 06:42:18.756894       1 statement.go:294] Remove Task <lm-mpi-job-mpiworker-0> on node <titanxp-4>
I1009 06:42:18.757029       1 allocate.go:153] Try to allocate resource to Jobs in Namespace <default> Queue <default>
I1009 06:42:18.757065       1 allocate.go:247] Leaving Allocate ...
I1009 06:42:18.757084       1 backfill.go:42] Enter Backfill ...
I1009 06:42:18.757116       1 backfill.go:72] Binding Task <default/lm-mpi-job-mpimaster-0> to node <titanxp-4>
I1009 06:42:18.757141       1 session.go:275] After allocated Task <default/lm-mpi-job-mpimaster-0> to Node <titanxp-4>: idle <cpu 31650.00, memory 270108176384.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00, nvidia.com/gpu 8000.00>, used <cpu 350.00, memory 31457280.00, nvidia.com/gpu 0.00>, releasing <cpu 0.00, memory 0.00>
I1009 06:42:18.757232       1 backfill.go:91] Leaving Backfill ...
I1009 06:42:18.762467       1 session.go:154] Close Session f0665e3c-ea5f-11e9-9d9a-e60f698a9621
```
The state of the job is
```
$ kubectl describe jobs.batch.volcano.sh lm-mpi-job 
Name:         lm-mpi-job
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"batch.volcano.sh/v1alpha1","kind":"Job","metadata":{"annotations":{},"name":"lm-mpi-job","namespace":"default"},"spec":{"mi...
API Version:  batch.volcano.sh/v1alpha1
Kind:         Job
Metadata:
  Creation Timestamp:  2019-10-09T06:41:00Z
  Generation:          1
  Resource Version:    13606927
  Self Link:           /apis/batch.volcano.sh/v1alpha1/namespaces/default/jobs/lm-mpi-job
  UID:                 30d14d7d-384c-4553-9239-4abb6dd7803e
Spec:
  Min Available:  3
  Plugins:
    Ssh:
    Svc:
  Queue:           default
  Scheduler Name:  volcano
  Tasks:
    Name:  mpimaster
    Policies:
      Action:  CompleteJob
      Event:   TaskCompleted
    Replicas:  1
    Template:
      Spec:
        Containers:
          Command:
            /bin/sh
            -c
            MPI_HOST=`cat /etc/volcano/mpiworker.host | tr "\n" ","`;
mkdir -p /var/run/sshd; /usr/sbin/sshd;
mpiexec --allow-run-as-root --host ${MPI_HOST} -np 2 mpi_hello_world > /home/re;

          Image:  volcanosh/example-mpi:0.0.1
          Name:   mpimaster
          Ports:
            Container Port:  22
            Name:            mpijob-port
          Working Dir:       /home
        Restart Policy:      OnFailure
    Name:                    mpiworker
    Replicas:                2
    Template:
      Spec:
        Containers:
          Command:
            /bin/sh
            -c
            mkdir -p /var/run/sshd; /usr/sbin/sshd -D;

          Image:  volcanosh/example-mpi:0.0.1
          Name:   mpiworker
          Ports:
            Container Port:  22
            Name:            mpijob-port
          Resources:
            Limits:
              nvidia.com/gpu:  1
          Working Dir:         /home
        Restart Policy:        OnFailure
Status:
  Controlled Resources:
    Plugin - Ssh:  ssh
    Plugin - Svc:  svc
  Min Available:   3
  Pending:         3
  State:
    Last Transition Time:  2019-10-09T06:41:02Z
    Phase:                 Pending
Events:                    <none>
```

**Comment by lminzhw** [comment]
**TLDR**: I think add resource requirement into master pod may solve this problem.

Aha, Your log show what scheduler do in this session precisely.
```
I1009 06:42:18.755985       1 allocate.go:208] Binding Task <default/lm-mpi-job-mpiworker-0> to node <titanxp-4>
I1009 06:42:18.756589       1 allocate.go:208] Binding Task <default/lm-mpi-job-mpiworker-1> to node <titanxp-1>
I1009 06:42:18.756787       1 statement.go:244] Allocating operations ...
I1009 06:42:18.756814       1 statement.go:310] Discarding operations ...
I1009 06:42:18.757065       1 allocate.go:247] Leaving Allocate ...
I1009 06:42:18.757084       1 backfill.go:42] Enter Backfill ...
I1009 06:42:18.757116       1 backfill.go:72] Binding Task <default/lm-mpi-job-mpimaster-0> to node <titanxp-4>
```

There is transaction during pod allocating. And pod without resource requirement would be allocated in backfill action but not allocate action.

So in this case, that allocation transaction treat this job gang unsatisfied and discarded the result (1 Pending 2 Allocated). And the backfill result (1 Allocated 2 Pending) is also discarded.

Finally, all pod in job remain in Pending status.

I think that is a bug, but it could be evaded by simply adding pod resource requirement.

**Comment by nicklhy** [comment]
@lminzhw Just tried adding the resource requirement into master pod and the job was launched successfully. **But this would result one gpu device wasting.**

Another question, there is a parameter named "slotsPerWorker" in kubeflow/mpi-operator. Is there a similar one in volcano mpi-job ? If no, how should I let the program know that I want to use multiple gpus on each machine instead of just multiple gpus over all available nodes? 

For instance, if I decided to launch a 64 gpu training job. I do not want the pods being scheduled on 64 nodes. It should be scheduled on exactly 8 nodes (if there are 8 gpus on each machine).

**Comment by lminzhw** [comment]
@nicklhy 
### Q1
That resource requirement is just a placeholder and adding `cpu: 1` may be enough.
### Q2
xref: https://www.open-mpi.org/faq/?category=running#mpirun-scheduling

The hostfile and `--host` arguments have its own method to describe multi-slot for a node. 

To describe node have 2 slots using hostfile
```
node0 slots=2 max_slots=20
node1 slots=2 max_slots=20
```

To describe node0 have 4 slots using `--host` arguments.
```
mpirun --host node0,node0,node0,node0 ....
```

So, for your 8 gpu machine, the hostfile may be like this
```
gpunode0 slots=8 .....
```

**Comment by lminzhw** [comment]
That `xref` is the config for `mpirun` or `mpiexec` command.

On Pod level, you can use 8 pod requires 8 gpu to gurantee that pod will occupy the whole machine and be scheduled on exactly 8 nodes finally. (may be hard to scheduled when cluster is occupied by others)

Or use 64 pod requires 1 gpu and use scheduler `binpack` policy to schedule them into same node. (may be not exactly 8 node occupied, it may run slowly but will run to complete finally)

**Comment by k82cn** [comment]
> The hostfile and --host arguments have its own method to describe multi-slot for a node.

We may consider to add a plugin in job controller for that to make it easy to use.

**Comment by nicklhy** [comment]
@lminzhw Thank you so much for the reply. But I am still a little bit confused for Q2.

I know how to do multi-gpu scheduling just like you mentioned above when manually executing mpirun commands. But when writing the job yaml files, the original command is
```shell
mpirun --allow-run-as-root -np "4" --host ${MPI_HOST} ...
```
Do I need to explicitly set the host names here?
```
mpirun --allow-run-as-root -np "4" --host node0,node0,node0,node0 ...
```
Notice the hostfile is not written by volcano users. We should be able to set multi-slot in the yaml file, right ?

**Comment by lminzhw** [comment]
@nicklhy 

As mentioned by @k82cn  , we may add this behavior into plugin later. 

But now, we could still achive this goal by using command like this.

```
MPI_HOST=`cat /etc/volcano/mpiworker.host | tr "\n" ","`;
mpiexec --allow-run-as-root --host ${MPI_HOST},${MPI_HOST} -np 2 mpi_hello_world;
```

Every host in MPI_HOST would be treat as having 2 slots by using this variable twice.

So, pod with 8 slot may represent by `--host ${MPI_HOST},${MPI_HOST},${MPI_HOST},${MPI_HOST},${MPI_HOST},${MPI_HOST},${MPI_HOST},${MPI_HOST}`. 

Aha, this command seems like a joke but would have right behavior.

Or, change the command to use hostfile with slots.
```
sed 's/$/& slots=8/' /etc/volcano/mpiworker.host > /tmp/hostfile
mpiexec --allow-run-as-root --hostfile /tmp/hostfile -np 2 mpi_hello_world;
```

Unfortunately, I have no environment to check all this command, but I think it may work. :)

**Comment by nicklhy** [comment]
@lminzhw Just verified your advice and it runs well. Really appreciate your valuable help !

I am also looking forward to your new plugin that implements this feature. The solution right now is not cool at all hahaha.

**Comment by nvcastet** [comment]
@lminzhw For your comment at https://github.com/volcano-sh/volcano/issues/474#issuecomment-539879519, could we create an issue/fix to allow the master pod not to have the `Resources` requirement?

**Comment by k82cn** [comment]
> For your comment at #474 (comment), could we create an issue/fix to allow the master pod not to have the Resources requirement?

That makes sence  to me :)

**Comment by k82cn** [comment]
@nicklhy , hm... I'm re-checking this thread; I have a question that: how many processes are you going to run in one single Pod?

**Comment by nicklhy** [comment]
@k82cn I was running distributed model training jobs on some 8 GPU servers. When using mpi-job, I guess there should be 8 processes in one single Pod since horovod usually launch a distinct process for each gpu. But when using mxnet-job, there should be only one process in one Pod.

**Comment by k82cn** [comment]
> since horovod usually launch a distinct process for each gpu

Does that mean horovod can only launch process on one GPU? Let's check this with horovod community :)
